src/Val.cc

```diff
@@ -4262,9 +4262,77 @@ ValPtr cast_value_to_type(Val* v, Type* t)
 		return static_cast<Broker::detail::DataVal*>(dv.get())->castTo(t);
 		}
 
+	// Allow casting between sets and vectors if the yield types are the same.
+	if ( v->GetType()->IsSet() && IsVector(t->Tag()) )
+		{
+		auto set_type = v->GetType<SetType>();
+		auto indices = set_type->GetIndices();
+
+		if ( indices->GetTypes().size() > 1 || ! indices->IsPure() )
+			return nullptr;
+
+		auto ret_type = make_intrusive<VectorType>(indices->GetPureType());
+		auto ret = make_intrusive<VectorVal>(ret_type);
+
+		auto lv = v->AsTableVal()->ToPureListVal();
+		auto n = lv->Length();
+
+		for ( int i = 0; i < n; ++i )
+			ret->Assign(i, lv->Idx(i));
+
+		return ret;
+		}
+	else if ( IsVector(v->GetType()->Tag()) && t->IsSet() )
+		{
+		auto tlp = make_intrusive<TypeList>(v->GetType()->Yield());
+		tlp->Append(v->GetType()->Yield());
+
+		auto ret_type = make_intrusive<SetType>(tlp, nullptr);
+		auto ret = make_intrusive<TableVal>(ret_type);
+
+		auto vv = v->AsVectorVal();
+		size_t size = vv->Size();
+
+		for ( size_t i = 0; i < size; i++ )
+			{
+			auto ve = vv->ValAt(i);
+			ret->Assign(std::move(ve), nullptr);
+			}
+
+		return ret;
+		}
+
 	return nullptr;
 	}
 
+static bool can_cast_set_and_vector(const Type* t1, const Type* t2)
+	{
+	const TableType* st = nullptr;
+	const VectorType* vt = nullptr;
+
+	if ( t1->IsSet() && IsVector(t2->Tag()) )
+		{
+		st = static_cast<const TableType*>(t1);
+		vt = static_cast<const VectorType*>(t2);
```

Comment:
More commonly, `AsTableType()` and `AsVectorType()` are used. There are no `static_cast` for Table/Vector Type so far.